In [1]:
import numpy as np
import pandas as pd

In [2]:
msft = pd.read_csv('MSFT.csv')
goog = pd.read_csv('GOOG.csv')
aapl = pd.read_csv('AAPL.csv')
spy = pd.read_csv('SPY.csv')

In [3]:
full_histories = np.concatenate([pd.DataFrame(df['Adj Close']) for df in [msft, goog, aapl, spy]], axis=1)

In [12]:
yearly_periods = 12
histories = full_histories[::len(full_histories)//(10*yearly_periods),:]

money_row = np.ones(shape=(len(histories),1))
histories = np.concatenate((money_row, histories), axis=1)

In [13]:
money_row.shape, histories.shape

((126, 1), (126, 5))

In [35]:
yearly_periods = 52
new_money_added = 400*12
histories = full_histories[::len(full_histories)//(10*yearly_periods),:]

money_row = np.ones(shape=(len(histories),1))
histories = np.concatenate((money_row, histories), axis=1)

commission = 5.5
final_values = []

for trigger_ in range(1, 101):
    trigger = trigger_/100

    bank = 100000
    ownership = np.array([bank] + [0]*4)

    for step in range(len(histories)):
        values = np.multiply(ownership, histories[step,:])
        mean_value = np.mean(values)
        diffs = (values - mean_value)
        percen_diffs = diffs/mean_value
        purchase_spots = np.concatenate((np.zeros(1), np.abs(percen_diffs[1:]) >= trigger))

        rebalance_purchases = np.multiply(-np.round(np.divide(diffs, histories[step,:]), decimals=-1),
                                          purchase_spots)
        spent = np.dot(rebalance_purchases, histories[step,:]) + commission * np.sum(purchase_spots) - new_money_added/yearly_periods

        assert bank >= 0
        ownership = ownership - np.array([spent] + [0]*4) + rebalance_purchases
    final_values.append(np.dot(ownership, histories[-1, :]))
    print("Trigger:", trigger, 
          "Avg value:", np.round(np.mean(final_values)))

Trigger: 0.01 Avg value: 380934.0
Trigger: 0.02 Avg value: 383653.0
Trigger: 0.03 Avg value: 385036.0
Trigger: 0.04 Avg value: 386406.0
Trigger: 0.05 Avg value: 386946.0
Trigger: 0.06 Avg value: 387181.0
Trigger: 0.07 Avg value: 387408.0
Trigger: 0.08 Avg value: 386980.0
Trigger: 0.09 Avg value: 387198.0
Trigger: 0.1 Avg value: 387159.0
Trigger: 0.11 Avg value: 387497.0
Trigger: 0.12 Avg value: 387522.0
Trigger: 0.13 Avg value: 387445.0
Trigger: 0.14 Avg value: 388027.0
Trigger: 0.15 Avg value: 387761.0
Trigger: 0.16 Avg value: 387852.0
Trigger: 0.17 Avg value: 388326.0
Trigger: 0.18 Avg value: 388246.0
Trigger: 0.19 Avg value: 388141.0
Trigger: 0.2 Avg value: 388676.0
Trigger: 0.21 Avg value: 388787.0
Trigger: 0.22 Avg value: 388601.0
Trigger: 0.23 Avg value: 388334.0
Trigger: 0.24 Avg value: 387537.0
Trigger: 0.25 Avg value: 386895.0
Trigger: 0.26 Avg value: 386033.0
Trigger: 0.27 Avg value: 385450.0
Trigger: 0.28 Avg value: 385128.0
Trigger: 0.29 Avg value: 385483.0
Trigger: 0.3 Avg

Yearly -> Trigger: 0.04 Avg value: 399943.0
Quarterly -> Trigger: 0.08 Avg value: 382868.0
Bi-Monthly actions -> Trigger: 0.3 Avg value: 372920.0
Monthly actions -> Trigger: 0.17 Avg value: 375073.0
Bi-Weekly -> Trigger: 0.34 Avg value: 380624.0
Weekly -> Trigger: 0.21 Avg value: 388787.0
